### PR TITLE
KAFKA-9965/KAFKA-13303: RoundRobinPartitioner broken by KIP-480

### PR DESCRIPTION
**Pull request by jonmcewen**
RoundRobinPartitioner behaviour was broken by sticky partitioning (KIP-480).

This patch addresses the behavioural issue caused by the second call to `partition()` after `onNewBatch()`, in a predicatable and thread-safe manner.  

Unit tested by simulation of multiple threads producing to two topics with race conditions.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jonmcewen** [comment]
Just checked my commit history and I had an earlier implementation like 
```
    // decrement the count for the partition for which the batch was aborted, as the last message
    // wasn't sent yet and partition() will be called again
    Optional.ofNullable(topicCounterMap.get(topic)).ifPresent(counter -> counter.decrementAndGet());
```
I'm pretty sure this didn't behave as desired when under load, hence the queueing solution, but I can't remember the exact behavioural issue.

**Comment by jonmcewen** [comment]
@showuon please review, thanks

**Comment by cmccabe** [comment]
Seems like this is the same issue as https://github.com/apache/kafka/pull/8690 ?

**Comment by Serwios** [comment]
Hi, is there any update on this fix?

File: clients/src/main/java/org/apache/kafka/clients/producer/RoundRobinPartitioner.java [hunk 1]
```diff
@@ -18,59 +18,99 @@
...
     }
 
     private int nextValue(String topic) {
-        AtomicInteger counter = topicCounterMap.computeIfAbsent(topic, k -> {
-            return new AtomicInteger(0);
-        });
+        AtomicInteger counter =
+            topicCounterMap.computeIfAbsent(
+                topic,
+                k -> {
```
Comment (Serwios) [inline clients/src/main/java/org/apache/kafka/clients/producer/RoundRobinPartitioner.java:86]: Can be simplified to adjust readability

	private int nextValue(String topic) {
		AtomicInteger counter = topicCounterMap.computeIfAbsent(topic, k -> new AtomicInteger(0));
		return counter.getAndIncrement();
	}

**Comment by Serwios** [comment]
@jonmcewen Don't you mind if I take initiative to drive this issue? So, if you not gonna continue work on this can you please close PR and I create a new one based on your changes

**Comment by chia7712** [comment]
onNewBatch is removed from 4.0 so I'm going to close this PR

**Comment by filpano** [comment]
@chia7712 It seems rather user unfriendly to not backport a fix to the completely RoundRobinPartitioner which is currently simply broken for old clients...

**Comment by chia7712** [comment]
> It seems rather user unfriendly to not backport a fix to the completely RoundRobinPartitioner which is currently simply broken for old clients...

The active branch with `onNewBatch` is 3.9. Would it be helpful to you if we backport the fix to 3.9? If so, I will file a PR for it.

**Comment by filpano** [comment]
Thanks for the earnest reply despite my somewhat tongue-in-cheek comment.

To shed some background light on the motivation: we currently have a use case where any large variation of record counts between partitions makes stream processing downstream prohibitively expensive due to required window sizes. As such, using something like the `RoundRobinPartitioner` seemed like the perfect fit to prevent skew in downstream processing tasks.

To put it another way: yes, a backport would be very helpful. 

For what it's worth, I've tested the updated partitioner logic here with a large number of records in a local test and it seems to be performing as it should (variance between partitions of _< 0.01%_).

If there's anything I can do in the way of further testing to help, please let me know.

**Comment by chia7712** [comment]
@filpano the bug in 3.9 was fixed by this commit: https://github.com/apache/kafka/commit/f670e9baa307d1b0abde871de17153a00d1fdd0a

**Comment by tanvp112** [comment]
@chia7712 [KAFKA-17632](https://issues.apache.org/jira/browse/KAFKA-17632) is not on the 3.9 release note. The 3.9.1 image was built ~4 months ago. Are we expecting 3.9.2 for this fix anytime soon?

**Comment by chia7712** [comment]
>  Are we expecting 3.9.2 for this fix anytime soon?

There is no thread in the dev channel for 3.9.2 yet. Have you considered upgrading to 4.x?

**Comment by tanvp112** [comment]
Can't do that until all other components, eg. apicurio, debezium & a host of many internal apps are fully tested & ready for 4.x. Would be very helpful if 3.9.x is keep updating while we plan & test 4.x.

**Comment by chia7712** [comment]
I have posted the discussion thread for 3.9.2 - https://lists.apache.org/thread/mz7nqbfl2ww7jvcko6qnoqow77xkly0d

**Comment by tanvp112** [comment]
@chia7712, greatly appreciate for the initiative! Hopefully it will get thru.
